File: Sources/Waldo/SteganographyEngine.swift
import CoreGraphics
import Foundation

class SteganographyEngine {
    
    static func embedWatermark(in image: CGImage, watermark: String, userID: String) -> CGImage? {
        let fullWatermark = "\(userID):\(watermark):\(Date().timeIntervalSince1970)"
        let watermarkData = Data(fullWatermark.utf8)
        
        guard let pixelData = extractPixelData(from: image) else { return nil }
        
        var modifiedPixelData = pixelData
        let totalPixels = image.width * image.height
        let watermarkBits = watermarkData.flatMap { byte in
            (0..<8).map { (byte >> $0) & 1 }
        }
        
        let lengthBits = withUnsafeBytes(of: UInt32(watermarkBits.count).bigEndian) { bytes in
            bytes.flatMap { byte in
                (0..<8).map { (byte >> $0) & 1 }
            }
        }
        
        let allBits = lengthBits + watermarkBits
        
        guard allBits.count <= totalPixels * 3 else { return nil }
        
        for (index, bit) in allBits.enumerated() {
            let pixelIndex = (index / 3) * 4 + (index % 3)
            if pixelIndex < modifiedPixelData.count {
                modifiedPixelData[pixelIndex] = (modifiedPixelData[pixelIndex] & 0xFE) | UInt8(bit)
            }
        }
        
        return createImage(from: modifiedPixelData, width: image.width, height: image.height)
    }
    
    static func extractWatermark(from image: CGImage) -> (userID: String, watermark: String, timestamp: TimeInterval)? {
        guard let pixelData = extractPixelData(from: image) else { return nil }
        
        let lengthBits = (0..<32).compactMap { bitIndex -> UInt8? in
            let pixelIndex = (bitIndex / 3) * 4 + (bitIndex % 3)
            guard pixelIndex < pixelData.count else { return nil }
            return pixelData[pixelIndex] & 1
        }
        
        let lengthBytes = stride(from: 0, to: lengthBits.count, by: 8).map { startIndex in
            lengthBits[startIndex..<min(startIndex + 8, lengthBits.count)]
                .enumerated()
                .reduce(UInt8(0)) { result, element in
                    result | (element.element << element.offset)
                }
        }
        
        let watermarkLength = Int(Data(lengthBytes).withUnsafeBytes { $0.load(as: UInt32.self).bigEndian })
        
        guard watermarkLength > 0 && watermarkLength <= 10000 else { return nil }
        
        let watermarkBits = (32..<32 + watermarkLength).compactMap { bitIndex -> UInt8? in
            let pixelIndex = (bitIndex / 3) * 4 + (bitIndex % 3)
            guard pixelIndex < pixelData.count else { return nil }
            return pixelData[pixelIndex] & 1
        }
        
        let watermarkBytes = stride(from: 0, to: watermarkBits.count, by: 8).map { startIndex in
            watermarkBits[startIndex..<min(startIndex + 8, watermarkBits.count)]
                .enumerated()
                .reduce(UInt8(0)) { result, element in
                    result | (element.element << element.offset)
                }
        }
        
        guard let watermarkString = String(data: Data(watermarkBytes), encoding: .utf8) else {
            return nil
        }
        
        let components = watermarkString.components(separatedBy: ":")
        guard components.count >= 3,
              let timestamp = TimeInterval(components.last!) else {
            return nil
        }
        
        let userID = components[0]
        let watermark = components[1..<components.count-1].joined(separator: ":")
        
        return (userID: userID, watermark: watermark, timestamp: timestamp)
    }
    
    private static func extractPixelData(from image: CGImage) -> [UInt8]? {
        let width = image.width
        let height = image.height
        let bytesPerPixel = 4
        let bytesPerRow = width * bytesPerPixel
        let totalBytes = height * bytesPerRow
        
        var pixelData = [UInt8](repeating: 0, count: totalBytes)
        
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let context = CGContext(
            data: &pixelData,
            width: width,
            height: height,
            bitsPerComponent: 8,
            bytesPerRow: bytesPerRow,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
        )
        
        guard let ctx = context else { return nil }
        
        ctx.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
        
        return pixelData
    }
    
    private static func createImage(from pixelData: [UInt8], width: Int, height: Int) -> CGImage? {
        let bytesPerPixel = 4
        let bytesPerRow = width * bytesPerPixel
        
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        
        var mutablePixelData = pixelData
        let context = mutablePixelData.withUnsafeMutableBytes { bytes in
            CGContext(
                data: bytes.baseAddress,
                width: width,
                height: height,
                bitsPerComponent: 8,
                bytesPerRow: bytesPerRow,
                space: colorSpace,
                bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
            )
        }
        
        return context?.makeImage()
    }
}

File: windows/src/Waldo.Core/SteganographyEngine.cs
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;
using SixLabors.ImageSharp.Processing;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Waldo.Core
{
    public class SteganographyEngine
    {
        public static Image<Rgba32>? EmbedWatermark(Image<Rgba32> image, string watermark, string userID)
        {
            string fullWatermark = $"{userID}:{watermark}:{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}";
            byte[] watermarkData = Encoding.UTF8.GetBytes(fullWatermark);
            
            var pixelData = ExtractPixelData(image);
            if (pixelData == null) return null;
            
            var modifiedPixelData = new byte[pixelData.Length];
            Array.Copy(pixelData, modifiedPixelData, pixelData.Length);
            
            int totalPixels = image.Width * image.Height;
            var watermarkBits = watermarkData.SelectMany(b => 
                Enumerable.Range(0, 8).Select(i => (byte)((b >> i) & 1))
            ).ToArray();
            
            var lengthBits = BitConverter.GetBytes((uint)watermarkBits.Length)
                .Reverse() // Big endian
                .SelectMany(b => Enumerable.Range(0, 8).Select(i => (byte)((b >> i) & 1)))
                .ToArray();
            
            var allBits = lengthBits.Concat(watermarkBits).ToArray();
            
            if (allBits.Length > totalPixels * 3) return null;
            
            for (int i = 0; i < allBits.Length; i++)
            {
                int pixelIndex = (i / 3) * 4 + (i % 3);
                if (pixelIndex < modifiedPixelData.Length)
                {
                    modifiedPixelData[pixelIndex] = (byte)((modifiedPixelData[pixelIndex] & 0xFE) | allBits[i]);
                }
            }
            
            return CreateImage(modifiedPixelData, image.Width, image.Height);
        }
        
        public static (string userID, string watermark, long timestamp)? ExtractWatermark(Image<Rgba32> image)
        {
            var pixelData = ExtractPixelData(image);
            if (pixelData == null) return null;
            
            var lengthBits = new byte[32];
            for (int i = 0; i < 32; i++)
            {
                int pixelIndex = (i / 3) * 4 + (i % 3);
                if (pixelIndex >= pixelData.Length) return null;
                lengthBits[i] = (byte)(pixelData[pixelIndex] & 1);
            }
            
            var lengthBytes = new byte[4];
            for (int i = 0; i < 4; i++)
            {
                byte value = 0;
                for (int j = 0; j < 8; j++)
                {
                    value |= (byte)(lengthBits[i * 8 + j] << j);
                }
                lengthBytes[i] = value;
            }
            
            Array.Reverse(lengthBytes); // Convert from big endian
            int watermarkLength = BitConverter.ToInt32(lengthBytes, 0);
            
            if (watermarkLength <= 0 || watermarkLength > 10000) return null;
            
            var watermarkBits = new byte[watermarkLength];
            for (int i = 0; i < watermarkLength; i++)
            {
                int bitIndex = 32 + i;
                int pixelIndex = (bitIndex / 3) * 4 + (bitIndex % 3);
                if (pixelIndex >= pixelData.Length) return null;
                watermarkBits[i] = (byte)(pixelData[pixelIndex] & 1);
            }
            
            var watermarkBytes = new byte[watermarkLength / 8];
            for (int i = 0; i < watermarkBytes.Length; i++)
            {
                byte value = 0;
                for (int j = 0; j < 8; j++)
                {
                    if (i * 8 + j < watermarkBits.Length)
                    {
                        value |= (byte)(watermarkBits[i * 8 + j] << j);
                    }
                }
                watermarkBytes[i] = value;
            }
            
            string watermarkString = Encoding.UTF8.GetString(watermarkBytes);
            if (string.IsNullOrEmpty(watermarkString)) return null;
            
            var components = watermarkString.Split(':');
            if (components.Length < 3) return null;
            
            if (!long.TryParse(components[^1], out long timestamp)) return null;
            
            string userID = components[0];
            string watermark = string.Join(":", components[1..^1]);
            
            return (userID, watermark, timestamp);
        }
        
        private static byte[]? ExtractPixelData(Image<Rgba32> image)
        {
            var pixelData = new byte[image.Width * image.Height * 4];
            
            image.ProcessPixelRows(accessor =>
            {
                int index = 0;
                for (int y = 0; y < accessor.Height; y++)
                {
                    var row = accessor.GetRowSpan(y);
                    for (int x = 0; x < row.Length; x++)
                    {
                        var pixel = row[x];
                        pixelData[index++] = pixel.R;
                        pixelData[index++] = pixel.G;
                        pixelData[index++] = pixel.B;
                        pixelData[index++] = pixel.A;
                    }
                }
            });
            
            return pixelData;
        }
        
        private static Image<Rgba32>? CreateImage(byte[] pixelData, int width, int height)
        {
            var image = new Image<Rgba32>(width, height);
            
            image.ProcessPixelRows(accessor =>
            {
                int index = 0;
                for (int y = 0; y < accessor.Height; y++)
                {
                    var row = accessor.GetRowSpan(y);
                    for (int x = 0; x < row.Length; x++)
                    {
                        if (index + 3 < pixelData.Length)
                        {
                            row[x] = new Rgba32(
                                pixelData[index++],
                                pixelData[index++],
                                pixelData[index++],
                                pixelData[index++]
                            );
                        }
                    }
                }
            });
            
            return image;
        }
    }
}